### PR TITLE
Webdriver

### PR DESCRIPTION
**Pull request by sibirtsv**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+def condition(driver):
+    asp_1 = aspect_ratio(driver, "#TOP_LEADERBOARD", "1")
+    return asp_1 == 2.4572192513368982
+
+
+# Start
+
+def main():
+    # Start options
+    server = Server(r"c:\browsermob\bin\browsermob-proxy.bat")
```
Comment (drets) [inline main.py]: Move string to constant at the top of file:
`PATH_TO_BROWSER_MOB = "c:\browsermob\bin\browsermob-proxy.bat"`

File: main.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    server.start()
+    proxy = server.create_proxy()
+    proxy.new_har()
+
+    chrome_options = webdriver.ChromeOptions()
+    url = urlparse(proxy.proxy).path
+    chrome_options.add_argument('--proxy-server=%s' % url)
+    driver = webdriver.Chrome(chrome_options=chrome_options)
+    wait = WebDriverWait(driver, 20)
+    driver.get("http://project43.wikia.com/wiki/SyntheticTests/VUAP")
```
Comment (drets) [inline main.py]: Move string to constant at the top of file.

File: main.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    # Make 1st screenshot
+    screenshot(driver, '#TOP_LEADERBOARD', "newest_before.png")
```
Comment (drets) [inline main.py]: Move strings to constant at the top of file.

File: main.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+    # Calculate aspect ratio of 1st image
+    asp_1 = aspect_ratio(driver, "#TOP_LEADERBOARD", "1")
+    print("aspect_ratio_before ", asp_1)
+
+    # Cheking 1st har for video element
+    result_1 = "video/mp4" in str(proxy.har)
```
Comment (drets) [inline main.py]: Move string to constant at the top of file.

File: main.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    assert not result_1
+    print("har_1", result_1)
+
+    # Enter iframe ans start video
+    ifrm = driver.find_element_by_css_selector(
+        '#TOP_LEADERBOARD iframe[title="3rd party ad content"]')
```
Comment (drets) [inline main.py]: Move string to constant at the top of file.

File: main.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    driver.switch_to.frame(ifrm)
+    driver.find_element_by_id('button').click()
+
+    # Quit iframe
+    driver.switch_to_default_content()
+
+    # Cheking visibility of the objects
+    close_b = element_displayed(wait, ".close-ad", "visible_1")
```
Comment (drets) [inline main.py]: Move string to constant at the top of file.

File: main.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    volume_b = element_displayed(wait, ".speaker", "visible_2")
```
Comment (drets) [inline main.py]: Move string to constant at the top of file.

File: main.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    assert close_b
+    assert volume_b
+    print("element_displayed ", close_b)
+    print("element_displayed ", volume_b)
+
+    # Make screenshot of video
+    screenshot(driver, '#TOP_LEADERBOARD', "video_screen.png")
```
Comment (drets) [inline main.py]: Move strings to constant at the top of file.

File: main.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    assert result_2
+    print("har_2", result_2)
+
+    wait.until(condition)
+
+    # Make 2nd screenshot
+    screenshot(driver, '#TOP_LEADERBOARD', "newest_after.png")
+
+    # Compare 1st and 2nd screenshots
+    scrn_comp = compare_screenshot("newest_before.png", "newest_after.png")
```
Comment (drets) [inline main.py]: Move string to constant at the top of file.

File: main.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    asp_rat_numb = size_picture_number['width'] / size_picture_number['height']
+    return asp_rat_numb
+
+
+# example aspect_ratio("#TOP_LEADERBOARD", "1")
+# example aspect_ratio("#TOP_LEADERBOARD", "2")
+
+
+def element_displayed(wait, visible_element_by_css,
+                      name_of_element):
```
Comment (drets) [inline main.py]: no need to pass `name_of_element` as method parameter.

File: main.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    # Calculate aspect ratio of 1st image
+    asp_1 = aspect_ratio(driver, "#TOP_LEADERBOARD", "1")
+    print("aspect_ratio_before ", asp_1)
+
+    # Cheking 1st har for video element
+    result_1 = "video/mp4" in str(proxy.har)
+    assert not result_1
+    print("har_1", result_1)
+
+    # Enter iframe ans start video
```
Comment (drets) [inline main.py]: typo.

File: main.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    # Make screenshot of video
+    screenshot(driver, '#TOP_LEADERBOARD', "video_screen.png")
+
+    # Calculate aspect ratio of video screenshot
+    asp_video = aspect_ratio(driver, "#TOP_LEADERBOARD", "3")
+
+    print("aspect_ratio_of_video ", asp_video)
+    assert asp_video != asp_1
+
+    # Wait fot 2nd screenshot
```
Comment (drets) [inline main.py]: I'd say we're waiting for 2nd screenshot [here](https://github.com/sibirtsv/Beginning/pull/3/files?diff=unified#diff-5bc02cefb3ea9e27f1a6776eabd1935dR134).

File: main.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    wait.until(condition)
+
+    # Make 2nd screenshot
+    screenshot(driver, '#TOP_LEADERBOARD', "newest_after.png")
+
+    # Compare 1st and 2nd screenshots
+    scrn_comp = compare_screenshot("newest_before.png", "newest_after.png")
+    assert scrn_comp, "Screenshots are not equal"
+    print("compare_screenshots", scrn_comp)
+
```
Comment (drets) [inline main.py]: Please close driver and remove all resources which were created during the test.

File: main.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        (By.CSS_SELECTOR, visible_element_by_css)))
+    return name_of_element.is_displayed()
+
+
+# example element_displayed("#ca-ve-edit", "visible_1")
+# example element_displayed("#CategorySelectAdd", "visible_2")
+
+def condition(driver):
+    asp_1 = aspect_ratio(driver, "#TOP_LEADERBOARD", "1")
+    return asp_1 == 2.4572192513368982
```
Comment (drets) [inline main.py]: This should be fixed. `2.4572192513368982` may be changed to any value − so we need to pass some parameter in order to _un-hardcode_ it.
(ping me if you won't solve it by yourself)

**Comment by drets** [comment]
Don't forget to run `flake8` after your changes again.

File: main.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+from browsermobproxy import Server
+from urllib.parse import urlparse
+from PIL import Image
+import math
+import operator
+from functools import reduce
+from selenium.webdriver.support import expected_conditions as EC
+
+
+def screenshot(driver, css_selector, screenshot_name_png):
```
Comment (sibirtsv) [inline main.py]: +

File: main.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+    rms = math.sqrt(reduce(operator.add,
+                           map(lambda a, b: (a - b) ** 2, h1, h2)) / len(h1))
+    return rms == 0.0
+
+
+# example compare_screenshot("newest.png", "screen_before.png")
+
+
+def aspect_ratio(driver, picture_css, picture_number):
```
Comment (sibirtsv) [inline main.py]: +

File: main.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    image_picture_number = driver.find_element_by_css_selector(picture_css)
+    size_picture_number = image_picture_number.size
+    asp_rat_numb = size_picture_number['width'] / size_picture_number['height']
+    return asp_rat_numb
+
+
+# example aspect_ratio("#TOP_LEADERBOARD", "1")
+# example aspect_ratio("#TOP_LEADERBOARD", "2")
```
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 10 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+
+def element_displayed(wait, visible_element_by_css,
+                      name_of_element):
+    name_of_element = wait.until(EC.visibility_of_element_located(
+        (By.CSS_SELECTOR, visible_element_by_css)))
+    return name_of_element.is_displayed()
+
+
+# example element_displayed("#ca-ve-edit", "visible_1")
+# example element_displayed("#CategorySelectAdd", "visible_2")
```
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 14 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 1 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 2 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 3 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 4 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 11 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 12 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 6 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 7 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 13 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+PATH_TO_BROWSER_MOB = r"c:\browsermob\bin\browsermob-proxy.bat"
+css_selector = "#TOP_LEADERBOARD"
+har_search = "video/mp4"
+iframe_css = '#TOP_LEADERBOARD iframe[title="3rd party ad content"]'
+screen_name_1 = "newest_before.png"
+screen_name_2 = "newest_after.png"
+screen_name_vid = "video_screen.png"
+el_displayed_1 = ".close-ad"
+el_displayed_2 = ".speaker"
+asp_rate_css = "#TOP_LEADERBOARD"
```
Comment (drets) [inline main.py]: Write constants in the following format: `PATH_TO_BROWSER_MOB` 
(all chars are capital!)

File: main.py [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+driver_get_url = "http://project43.wikia.com/wiki/SyntheticTests/VUAP"
+PATH_TO_BROWSER_MOB = r"c:\browsermob\bin\browsermob-proxy.bat"
+css_selector = "#TOP_LEADERBOARD"
+har_search = "video/mp4"
+iframe_css = '#TOP_LEADERBOARD iframe[title="3rd party ad content"]'
+screen_name_1 = "newest_before.png"
+screen_name_2 = "newest_after.png"
+screen_name_vid = "video_screen.png"
+el_displayed_1 = ".close-ad"
+el_displayed_2 = ".speaker"
```
Comment (drets) [inline main.py]: Variable name should answer a question: “what is that?”
E.g. `el_displayed_2 = ".speaker"` should be `SPEAKER_CSS_SELECTOR = ".speaker"`

File: main.py [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+
+def element_displayed(wait, visible_element_by_css):
+    """Show visibility for chosen objects
+    example element_displayed("#ca-ve-edit", "visible_1")"""
+    visible_elem = wait.until(EC.visibility_of_element_located(
+        (By.CSS_SELECTOR, visible_element_by_css)))
+    return visible_elem.is_displayed()
+
+
+def condition(driver):
```
Comment (drets) [inline main.py]: Do we need this function?

File: main.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+
+    # Make 2nd screenshot
+    screenshot(driver, css_selector, screen_name_2)
+
+    # Compare 1st and 2nd screenshots
+    scrn_comp = compare_screenshot(screen_name_1, screen_name_2)
+    assert scrn_comp, "Screenshots are not equal"
+    print("compare_screenshots", scrn_comp)
+
+    server.stop()
```
Comment (drets) [inline main.py]: Move next 6 lines to separate function: 
e.g. `afterTest(server, driver, screen_name_1, screen_name_2)`)

File: main.py [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+    # Cheking visibility of the objects
+    close_b = element_displayed(wait, el_displayed_1)
+    volume_b = element_displayed(wait, el_displayed_2)
+    assert close_b
+    assert volume_b
+    print("element_displayed ", close_b)
+    print("element_displayed ", volume_b)
+
+    # Make screenshot of video
+    screenshot(driver, css_selector, screen_name_vid)
```
Comment (drets) [inline main.py]: We don't use screenshot of video therefore we don't need this line.

File: main.py (hunk 20 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 21 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py (hunk 22 above)
Comment (sibirtsv) [inline main.py]: +

File: main.py [hunk 24]
```diff
@@ -86,71 +79,63 @@ def main():
...
     assert scrn_comp, "Screenshots are not equal"
     print("compare_screenshots", scrn_comp)
 
-    server.stop()
-    driver.quit()
-
-    os.remove(screen_name_1)
-    os.remove(screen_name_2)
-    os.remove(screen_name_vid)
+    server.stop(), driver.quit(), os.remove(SCREEN_NAME_1), os.remove(SCREEN_NAME_2)
```
Comment (drets) [inline main.py]: It looks strange, could you move it to separate function and have it on 4 lines?

File: main.py [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+
+    # Make 2nd screenshot
+    screenshot(driver, SCREEN_CSS, SCREEN_NAME_2)
+
+    # Compare 1st and 2nd screenshots
+    scrn_comp = compare_screenshot(SCREEN_NAME_1, SCREEN_NAME_2)
+    assert scrn_comp, "Screenshots are not equal"
+    print("compare_screenshots", scrn_comp)
+
+    server.stop(), driver.quit(), os.remove(SCREEN_NAME_1), os.remove(SCREEN_NAME_2)
```
Comment (drets) [inline main.py]: `main.py:138:80: E501 line too long (84 > 79 characters)`
Please, run `flake8` after each change.

**Pull request merged**